File: multicontactpicker/src/main/java/com/wafflecopter/multicontactpicker/Views/RoundLetterView.java
package com.wafflecopter.multicontactpicker.Views;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.graphics.Typeface;
import android.text.TextPaint;
import android.util.AttributeSet;
import android.view.View;

import com.wafflecopter.multicontactpicker.R;

/*
*   Copyright 2014 Pavlos-Petros Tournaris
*
*   Licensed under the Apache License, Version 2.0 (the "License");
*   you may not use this file except in compliance with the License.
*   You may obtain a copy of the License at
*
*   http://www.apache.org/licenses/LICENSE-2.0
*
*   Unless required by applicable law or agreed to in writing, software
*   distributed under the License is distributed on an "AS IS" BASIS,
*   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*   See the License for the specific language governing permissions and
*   limitations under the License.
*/

public class RoundLetterView extends View {

    private static int DEFAULT_TITLE_COLOR = Color.WHITE;
    private static int DEFAULT_BACKGROUND_COLOR = Color.CYAN;
    private static final int DEFAULT_VIEW_SIZE = 96;
    private static float DEFAULT_TITLE_SIZE = 25f;
    private static String DEFAULT_TITLE = "A";

    private int mTitleColor = DEFAULT_TITLE_COLOR;
    private int mBackgroundColor = DEFAULT_BACKGROUND_COLOR;
    private String mTitleText = DEFAULT_TITLE;
    private float mTitleSize = DEFAULT_TITLE_SIZE;

    private TextPaint mTitleTextPaint;
    private Paint mBackgroundPaint;
    private RectF mInnerRectF;
    private int mViewSize;

    private Typeface mFont = Typeface.defaultFromStyle(Typeface.BOLD);

    public RoundLetterView(Context context) {
        super(context);
        init(null, 0);
    }

    public RoundLetterView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(attrs, 0);
    }

    public RoundLetterView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(attrs, defStyle);
    }

    private void init(AttributeSet attrs, int defStyle) {
        final TypedArray a = getContext().obtainStyledAttributes(
                attrs, R.styleable.RoundedLetterView, defStyle, 0);

        if(a.hasValue(R.styleable.RoundedLetterView_rlv_titleText)){
            mTitleText = a.getString(R.styleable.RoundedLetterView_rlv_titleText);
        }

        mTitleColor = a.getColor(R.styleable.RoundedLetterView_rlv_titleColor,DEFAULT_TITLE_COLOR);
        mBackgroundColor = a.getColor(R.styleable.RoundedLetterView_rlv_backgroundColorValue,DEFAULT_BACKGROUND_COLOR);

        mTitleSize = a.getDimension(R.styleable.RoundedLetterView_rlv_titleSize,DEFAULT_TITLE_SIZE);
        a.recycle();

        //Title TextPaint
        mTitleTextPaint = new TextPaint();
        mTitleTextPaint.setFlags(Paint.ANTI_ALIAS_FLAG);
        mTitleTextPaint.setTypeface(mFont);
        mTitleTextPaint.setTextAlign(Paint.Align.CENTER);
        mTitleTextPaint.setLinearText(true);
        mTitleTextPaint.setColor(mTitleColor);
        mTitleTextPaint.setTextSize(mTitleSize);

        //Background Paint
        mBackgroundPaint = new Paint();
        mBackgroundPaint.setFlags(Paint.ANTI_ALIAS_FLAG);
        mBackgroundPaint.setStyle(Paint.Style.FILL);
        mBackgroundPaint.setColor(mBackgroundColor);

        mInnerRectF = new RectF();
    }

    private void invalidateTextPaints(){
        mTitleTextPaint.setTypeface(mFont);
        mTitleTextPaint.setTextSize(mTitleSize);
        mTitleTextPaint.setColor(mTitleColor);
    }

    private void invalidatePaints(){
        mBackgroundPaint.setColor(mBackgroundColor);
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);

        int width = resolveSize(DEFAULT_VIEW_SIZE, widthMeasureSpec);
        int height = resolveSize(DEFAULT_VIEW_SIZE, heightMeasureSpec);
        mViewSize = Math.min(width, height);

        setMeasuredDimension(width, height);
    }

    @Override
    protected void onDraw(Canvas canvas) {

        mInnerRectF.set(0, 0, mViewSize, mViewSize);
        mInnerRectF.offset((getWidth() - mViewSize) / 2, (getHeight() - mViewSize) / 2);

        float centerX = mInnerRectF.centerX();
        float centerY = mInnerRectF.centerY();

        int xPos = (int) centerX;
        int yPos = (int) (centerY - (mTitleTextPaint.descent() + mTitleTextPaint.ascent()) / 2);

        canvas.drawOval(mInnerRectF, mBackgroundPaint);

        canvas.drawText(mTitleText,
                xPos,
                yPos,
                mTitleTextPaint);
    }

    /**
     * Gets the title string attribute value.
     * @return The title string attribute value.
     */
    public String getTitleText() {
        return mTitleText;
    }

    /**
     * Sets the view's title string attribute value.
     * @param title The example string attribute value to use.
     */
    public void setTitleText(String title) {
        mTitleText = title;
        invalidate();
    }

    /**
     * Gets the background color attribute value.
     * @return The background color attribute value.
     */
    public int getBackgroundColor() {
        return mBackgroundColor;
    }

    /**
     * Sets the view's background color attribute value.
     * @param backgroundColor The background color attribute value to use.
     */
    public void setBackgroundColor(int backgroundColor) {
        mBackgroundColor = backgroundColor;
        invalidatePaints();
    }

    /**
     * Gets the title size dimension attribute value.
     * @return The title size dimension attribute value.
     */
    public float getTitleSize() {
        return mTitleSize;
    }

    /**
     * Sets the view's title size dimension attribute value.
     * @param titleSize The title size dimension attribute value to use.
     */
    public void setTitleSize(float titleSize) {
        mTitleSize = titleSize;
        invalidateTextPaints();
    }

    /**
     * Sets the view's title typeface.
     * @param font The typeface to be used for the text.
     */
    public void setTextTypeface(Typeface font){
        this.mFont = font;
        invalidateTextPaints();
    }

    /**
     * Sets the view's title color attribute value.
     * @param titleColor The title color attribute value to use.
     */
    public void setTitleColor(int titleColor) {
        mTitleColor = titleColor;
        invalidateTextPaints();
    }
}

File: contactspicker/src/main/java/com/wafflecopter/contactspicker/Views/RoundLetterView.kt
package com.wafflecopter.contactspicker.Views

import android.content.Context
import android.graphics.*
import android.text.TextPaint
import android.util.AttributeSet
import android.view.View
import com.wafflecopter.contactspicker.R

/*
*   Copyright 2014 Pavlos-Petros Tournaris
*
*   Licensed under the Apache License, Version 2.0 (the "License");
*   you may not use this file except in compliance with the License.
*   You may obtain a copy of the License at
*
*   http://www.apache.org/licenses/LICENSE-2.0
*
*   Unless required by applicable law or agreed to in writing, software
*   distributed under the License is distributed on an "AS IS" BASIS,
*   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
*   See the License for the specific language governing permissions and
*   limitations under the License.
*/
class RoundLetterView : View {
    private var mTitleColor = DEFAULT_TITLE_COLOR
    private var mBackgroundColor = DEFAULT_BACKGROUND_COLOR
    private var mTitleText: String? = DEFAULT_TITLE
    private var mTitleSize = DEFAULT_TITLE_SIZE
    private var mTitleTextPaint: TextPaint? = null
    private var mBackgroundPaint: Paint? = null
    private var mInnerRectF: RectF? = null
    private var mViewSize = 0
    private var mFont = Typeface.defaultFromStyle(Typeface.BOLD)

    constructor(context: Context?) : super(context) {
        init(null, 0)
    }

    constructor(context: Context?, attrs: AttributeSet?) : super(context, attrs) {
        init(attrs, 0)
    }

    constructor(context: Context?, attrs: AttributeSet?, defStyle: Int) : super(
        context,
        attrs,
        defStyle
    ) {
        init(attrs, defStyle)
    }

    private fun init(attrs: AttributeSet?, defStyle: Int) {
        val a = context.obtainStyledAttributes(
            attrs, R.styleable.RoundLetterView, defStyle, 0
        )
        if (a.hasValue(R.styleable.RoundLetterView_rlv_titleText)) {
            mTitleText = a.getString(R.styleable.RoundLetterView_rlv_titleText)
        }
        mTitleColor = a.getColor(R.styleable.RoundLetterView_rlv_titleColor, DEFAULT_TITLE_COLOR)
        mBackgroundColor = a.getColor(
            R.styleable.RoundLetterView_rlv_backgroundColorValue,
            DEFAULT_BACKGROUND_COLOR
        )
        mTitleSize = a.getDimension(R.styleable.RoundLetterView_rlv_titleSize, DEFAULT_TITLE_SIZE)
        a.recycle()

        //Title TextPaint
        mTitleTextPaint = TextPaint()
        mTitleTextPaint!!.flags = Paint.ANTI_ALIAS_FLAG
        mTitleTextPaint!!.typeface = mFont
        mTitleTextPaint!!.textAlign = Paint.Align.CENTER
        mTitleTextPaint!!.isLinearText = true
        mTitleTextPaint!!.color = mTitleColor
        mTitleTextPaint!!.textSize = mTitleSize

        //Background Paint
        mBackgroundPaint = Paint()
        mBackgroundPaint!!.flags = Paint.ANTI_ALIAS_FLAG
        mBackgroundPaint!!.style = Paint.Style.FILL
        mBackgroundPaint!!.color = mBackgroundColor
        mInnerRectF = RectF()
    }

    private fun invalidateTextPaints() {
        mTitleTextPaint!!.typeface = mFont
        mTitleTextPaint!!.textSize = mTitleSize
        mTitleTextPaint!!.color = mTitleColor
    }

    private fun invalidatePaints() {
        mBackgroundPaint!!.color = mBackgroundColor
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)
        val width = resolveSize(DEFAULT_VIEW_SIZE, widthMeasureSpec)
        val height = resolveSize(DEFAULT_VIEW_SIZE, heightMeasureSpec)
        mViewSize = Math.min(width, height)
        setMeasuredDimension(width, height)
    }

    override fun onDraw(canvas: Canvas) {
        mInnerRectF!![0f, 0f, mViewSize.toFloat()] = mViewSize.toFloat()
        mInnerRectF!!.offset((width - mViewSize) / 2.toFloat(), (height - mViewSize) / 2.toFloat())
        val centerX = mInnerRectF!!.centerX()
        val centerY = mInnerRectF!!.centerY()
        val xPos = centerX.toInt()
        val yPos =
            (centerY - (mTitleTextPaint!!.descent() + mTitleTextPaint!!.ascent()) / 2).toInt()
        canvas.drawOval(mInnerRectF!!, mBackgroundPaint!!)
        canvas.drawText(
            mTitleText!!,
            xPos.toFloat(),
            yPos.toFloat(),
            mTitleTextPaint!!
        )
    }
    /**
     * Gets the title string attribute value.
     * @return The title string attribute value.
     */
    /**
     * Sets the view's title string attribute value.
     * @param titleText The example string attribute value to use.
     */
    var titleText: String?
        get() = mTitleText
        set(title) {
            mTitleText = title
            invalidate()
        }

    /**
     * Gets the background color attribute value.
     * @return The background color attribute value.
     */
    fun getBackgroundColor(): Int {
        return mBackgroundColor
    }

    /**
     * Sets the view's background color attribute value.
     * @param backgroundColor The background color attribute value to use.
     */
    override fun setBackgroundColor(backgroundColor: Int) {
        mBackgroundColor = backgroundColor
        invalidatePaints()
    }
    /**
     * Gets the title size dimension attribute value.
     * @return The title size dimension attribute value.
     */
    /**
     * Sets the view's title size dimension attribute value.
     * @param titleSize The title size dimension attribute value to use.
     */
    var titleSize: Float
        get() = mTitleSize
        set(titleSize) {
            mTitleSize = titleSize
            invalidateTextPaints()
        }

    /**
     * Sets the view's title typeface.
     * @param font The typeface to be used for the text.
     */
    fun setTextTypeface(font: Typeface) {
        mFont = font
        invalidateTextPaints()
    }

    /**
     * Sets the view's title color attribute value.
     * @param titleColor The title color attribute value to use.
     */
    fun setTitleColor(titleColor: Int) {
        mTitleColor = titleColor
        invalidateTextPaints()
    }

    companion object {
        private const val DEFAULT_TITLE_COLOR = Color.WHITE
        private const val DEFAULT_BACKGROUND_COLOR = Color.CYAN
        private const val DEFAULT_VIEW_SIZE = 96
        private const val DEFAULT_TITLE_SIZE = 25f
        private const val DEFAULT_TITLE = "A"
    }
}